Add changes to readme (#7298)
semgrep/semgrep: README.md

Semgrep is a fast, open-source, static analysis engine for finding bugs, detecting vulnerabilities in third-party dependencies, and enforcing code standards. Semgrep analyzes code locally on your computer or in your build environment: **code is never uploaded**. [Get started →.](#getting-started-🚀)

<img src="images/semgrep-scan-cli.jpg" alt="Semgrep CLI image"/>

### Language support

Semgrep supports 30+ languages.

| Category     | Languages                                                                                                                                                            |
| ------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| GA           | C# · Go · Java · JavaScript · JSX · JSON · PHP · Python · Ruby · Scala · Terraform · TypeScript · TSX                                                                |

> | Beta         | Kotlin . Rust                                                                                                                                                        |

Reviewer: ```suggestion
| Beta         | Kotlin · Rust                                                                                                                                                        |
```

> | Experimental | Bash . C . C++ . Clojure . Dart . Dockerfile . Elixir . HTML . Jsonnet . Lisp . Lua . OCaml . R . Scheme . Solidity . Swift . YAML . XML . Generic (ERB, Jinja, etc) |

Reviewer: ```suggestion
| Experimental | Bash · C · C++ · Clojure · Dart · Dockerfile · Elixir · HTML · Jsonnet · Lisp · Lua · OCaml · R · Scheme · Solidity · Swift · YAML · XML · Generic (ERB, Jinja, etc.) |
```

</p>
</br>

> Semgrep is a fast, open-source, static analysis engine for finding bugs, detecting vulnerabilities in third-party dependencies, and enforcing code standards. Semgrep analyzes code locally on your computer or in your build environment: **code is never uploaded**. [Get started →.](#getting-started)

Reviewer 2: The link here is broken, I believe the 🚀 emoji needs to be included.

Author: Fixed

> <img src="images/semgrep-main-image.jpg" alt="Semgrep platform image"/>

Reviewer 2: This image and coloration is quite lovely. wydt of us leading with a terminal output image since this is the home of our OSS project? Save this photo as a value-pitch in the ui section?

Author: Thanks. Our terminal output isn't good just yet, but once we have it, we can certainly come back to this.

Semgrep is a fast, open-source, static analysis engine for finding bugs, detecting vulnerabilities in third-party dependencies, and enforcing code standards. [Get started →.](#getting-started)

> <img src="images/semgrep-main-image.jpg" />

Reviewer: @rajaraodv Can you remove the drop shadow, blue border, and curvature on this image? I think it would look better if it’s a plain screenshot, and the same for other images further down in the README.

Also, can you change the paths of the image sources to match the convention used by the logo on line 3? This is needed because the README is automatically displayed in other places (like PyPI) and those need an absolute URL to render the image.

Author: Got it!

<details>

>   <summary style="font-size:16px;">Semgrep Ecosystem</summary>

Reviewer 2: @pabloest, is it critical that we include information in this README about Semgrep Cloud Platform, Code, and Supply Chain, or could we link instead to Semgrep.dev for that, with something along the lines of "Learn more about Semgrep Ecosystem"?

Author: For now, it's hidden under the details toggle and so it can remain there.

Reviewer: We added them to the README because we thought it is critical to include, and I don’t think we have updated information to show otherwise, so we should leave it in there. I think having it under the toggle dropdown might be ok, but I ask myself why it should be default hidden given its placement in the order of the README? 

Assuming many readers don't click on the toggle dropdown (it’s easy to miss), then there is much less context shown to the reader about what is Semgrep aside from the single opening sentence.

Author: discussed this with @pabloest and have moved it to after the Getting started section to keep the focus getting started.

> Semgrep analyzes code locally on your computer or in your build environment: **code is never uploaded**.

Reviewer 2: @rajaraodv, this content has been important historically for our OSS users, up to the point after the image. In particular, there are continued concerns that Semgrep sends code to the cloud, so we've made a point of highlighting it here. 

My suggestion for the details section is to only capture the ecosystem bullet points below, and not this Semgrep OSS content.

Author: I have moved the details section to the only capture the ecosystem part.

<p align="center">
See <a href="https://semgrep.dev/docs/supported-languages/">supported languages</a> for the complete list.
</p>
<br/>

### Getting started 🚀

> 1. [Getting started from the CLI](#option-1-getting-started-from-the-cli)
> 2. [Getting started from the web UI](#option-2-getting-started-via-the-web-ui-recommended)

Reviewer 2: ```suggestion
1. [Getting started (no login)](#option-1-getting-started-from-the-cli)
2. [Getting started (Semgrep Cloud Platform)](#option-2-getting-started-via-the-web-ui-recommended)
```

> We recommend starting from the [web UI](#option-2-getting-started-via-the-web-ui-recommended) and then install the CLI because it is simple and allows you to quickly see and manage results, as well as set up CI/CD pipelines with minimal effort. You can also use additional proprietary Semgrep Registry rules and run custom policies from the Semgrep App.
> 
> However, you can also start with the CLI directly, even without logging in, but you will lose the benefits mentioned earlier.

Reviewer 2: ```suggestion
We recommend starting with the [Cloud Platform](#option-2-getting-started-via-the-web-ui-recommended) because it provides a demo project, result triaging and exploration workflows, and makes set up in CI/CD fast. Scans are still local and code isn't uploaded. Alternatively start without logging in, and navigate terminal output and run one-off searches.
```

> ### Option 1: Getting started from the CLI

Reviewer 2: alt:

### Option 1: Getting started (no login)

1.  Install Semgrep CLI

```
# For macOS
$ brew install semgrep

# For Ubuntu/WSL/Linux/macOS
$ python3 -m pip install semgrep

# To try Semgrep without installation run via Docker
$ docker run --rm -v "${PWD}:/src" returntocorp/semgrep semgrep
```

2. Go to a repository's root directory and run `semgrep --config auto`. This will scan your project with default settings.

  ```

3.  Run `semgrep login` to get the login URL. Open the login URL in the browser and login.

<img src="images/semgrep-login.jpg" /> 

4. Go to your app's root directory and run `semgrep --config auto`

<img src="images/semgrep-auto-config-example.jpg" /> 

### Getting started (CLI only)
In this flow, you'll lose the simplicity of quickly seeing and managing results and setting up CI/CD pipelines with minimal effort. Additionally, you'll be unable to use additional proprietary Semgrep Registry rules and run custom policies from the Semgrep App.

 1. Install Semgrep CLI

  ```
  # For macOS
  $ brew install semgrep

  # For Ubuntu/WSL/Linux/macOS
  $ python3 -m pip install semgrep

  # To try Semgrep without installation run via Docker
  $ docker run --rm -v "${PWD}:/src" returntocorp/semgrep semgrep
  ```

2. Go to your app's root directory and run the following:
```
semgrep --config auto
```

### Set up CI/CD in two minutes

1.  Register or login to <a href="http://bit.ly/424Jbs0" target="_blank">semgrep.dev</a>

2. Go to `Projects > Scan New Project > Run scan in CI`
<img src="images/semgrep-ci-cd-popup.jpg" /> 

> 3. Select your version control system and follow the wizard to add your project. After this setup, you'll get continuous security. Semgrep will scan your project after every pull request!

Reviewer 3: I'm not sure about this line: "you'll get continuous security" (our rules / support vary per language) - I wonder if there's a more accurate way to explain the value add?

### Getting started

> 1.  Register to <a href="http://bit.ly/424Jbs0" target="_blank">semgrep.dev</a>

Reviewer 2: @rajaraodv, what if we maintain the old sequence, principally that installation is step #1? This helps both to start quickly and aligns with our FOSS philosophy. After that, suggest login and clearly articulate the value-add for people using Semgrep CLI.

> 1.  Register to <a href="http://bit.ly/424Jbs0" target="_blank">semgrep.dev</a>

Reviewer 2: The use of a bit.ly link here feels out of place. As a security professional I am hesitant to click on such links, and would prefer to either see a semgrep.dev domain or all of the details of the utm/campaign.

<img src="images/semgrep-main-image.jpg" /> 

2.  Install Semgrep CLI

>   ```

Reviewer 2: For styling, set the language of the block to sh: ```sh, this should be applied across code blocks.

  # For macOS
  $ brew install semgrep

  # For Ubuntu/WSL/Linux/macOS
  $ python3 -m pip install semgrep

  # To try Semgrep without installation run via Docker
  $ docker run --rm -v "${PWD}:/src" returntocorp/semgrep semgrep
  ```

3.  Run `semgrep login` to get the login URL. Open the login URL in the browser and login.

>   ```

Reviewer 2: For styling, set the language of the block to sh: ```sh, this should be applied across code blocks.

>   semgrep login

Reviewer 2: We preface commands with `$`, looking at previous blocks

  ```
<img src="images/semgrep-login.jpg" /> 

4. Go to your app's root directory and run the following:

```
semgrep --config auto
```
<img src="images/semgrep-auto-config-example.jpg" /> 

> ### Set up CI/CD in two minutes

Reviewer 2: My suggestion is to keep the existing documentation, but add a new block about Semgrep.dev login and connected scans here. This matches what I see in sibling tool READMEs, allowing us to highlight the value of going deeper into the Semgrep ecosystem.

<p align="center">
See <a href="https://semgrep.dev/docs/supported-languages/">supported languages</a> for the complete list.
</p>
<br/>

### Getting started
There are two ways to get started. 

1. Getting started from the CLI. 
2. Getting started from the web UI.

We recommend starting from the web interface and then install the CLI because it is simple and allows you to quickly see and manage results, as well as set up CI/CD pipelines with minimal effort. You can also use additional proprietary Semgrep Registry rules and run custom policies from the Semgrep App. 

However, you can also start with the CLI directly, even without logging in, but you will lose the benefits mentioned earlier.

### Option 1: Getting started from the CLI

> 1.  Install Semgrep CLI

Reviewer 2: I think this is the same first step for both "paths" and we can/should lead with it. Once installed, _then_ people have the option of `semgrep login` to get all the features of Semgrep Cloud Platform, Code, and Supply Chain (value add), or using Semgrep OSS logged-out. 

In both cases, the output of the scan will look the same, but you'll have the option in the authenticated use case to triage and review issues in the web interface (note: only results metadata is sent, code is not uploaded, see privacy.md)

Author: I'm not entirely sure about that. Personally, when I used Semgrep, I didn't even need to download the CLI. The web interface was sufficient for me to identify all the issues and manage them effectively. Additionally, our CI/CD setup will handle scanning every time I submit a PR. So, in this context, using the CLI might not be necessary. I feel like we get more value and a better "ah ha" moment with the web flow.

$ docker run --rm -v "${PWD}:/src" returntocorp/semgrep semgrep
```

> 2.  Run `semgrep login` to get the login URL. Open the login URL in the browser and login [Optional, but recommended].

Reviewer 4: You don't need to open the login url. `semgrep login` should be sufficient and should open the browser automatically

Author: It never opened the browser for me. I thought it was not implemented.

3.  Go to your app's root directory and run `semgrep --config auto`. This will scan your project with the default settings.

### Option 2: Getting started via the web UI (Recommended)

> 1.  Register to <a href="https://go.semgrep.dev/login-ghrmgo" target="_blank">semgrep.dev</a>

Reviewer 2: ```
### Option 2: Getting started (Semgrep Cloud Platform)

1.  Log in to <a href="https://go.semgrep.dev/login-ghrmgo" target="_blank">Semgrep Cloud Platform</a>
```

$ docker run --rm -v "${PWD}:/src" returntocorp/semgrep semgrep
```

2.  Run `semgrep login` to get the login URL. Open the login URL in the browser and login [Optional, but recommended].

3.  Go to your app's root directory and run `semgrep --config auto`. This will scan your project with the default settings.

### Option 2: Getting started via the web UI (Recommended)

1.  Register to <a href="https://semgrep.dev/login?utm_source=github&utm_medium=readme&utm_campaign=growth-oss" target="_blank">semgrep.dev</a>

2.  Add your project for scanning by go to `Projects > Scan New Project > Run scan in CI`
    <img src="images/semgrep-ci-cd-popup.jpg" />

3.  Select your version control system and follow the wizard to add your project. After this setup, Semgrep will scan your project after every pull request. You don't really need the CLI unless you want to scan on local machine.

### Notes:

1.  Visit [Docs > Running rules](https://semgrep.dev/docs/running-rules/) to learn more about `auto` config and other rules.

> 2.  If there are any issues, please ask us at our Slack channel <a href="https://r2c.dev/slack"> https://r2c.dev/slack</a>

Reviewer: ```suggestion
2.  If there are any issues, please ask us at our Slack channel <a href="https://go.semgrep.dev/slack"> https://go.semgrep.dev/slack</a>
```
